extern/storage-sealing/upgrade_queue.go

```diff
@@ -44,11 +77,38 @@ func (m *Sealing) MarkForUpgrade(id abi.SectorNumber) error {
 		return xerrors.Errorf("not a committed-capacity sector, has deals")
 	}
 
-	// TODO: more checks to match actor constraints
+	tok, head, err := m.Api.ChainHead(ctx)
+	if err != nil {
+		return xerrors.Errorf("couldnt get chain head: %w", err)
+	}
+	onChainInfo, err := m.Api.StateSectorGetInfo(ctx, m.maddr, id, tok)
+	if err != nil {
+		return xerrors.Errorf("failed to read sector on chain info: %w", err)
+	}
 
-	m.toUpgrade[id] = struct{}{}
+	active, err := m.Api.StateMinerActiveSectors(ctx, m.maddr, tok)
```

Comment:
That's going to be a bit O(nSectors^2)

Reply:
(but probably not worth optimizing now given that it will be called once per 32-64G of data)

Reply:
Any existing api calls I can use that don't go through all sectors?

Reply:
We could load the miner state, and use the bitfields